### ISSUE TITLE
to_csv chokes if not passed sep as a string, even when encoding is set to unicode

### ISSUE THREAD

**Issue by maxgrenderjones** [pydata/pandas#6035]
I am using anaconda python 2.7, pandas 0.12.0, with `from __future__ import unicode_literals`

#### The bug

When I call `dataframe.to_csv(fname, sep=sep, index=index, float_format=float_format, **csvkwargs)` and pass in `sep=','`, pandas throws an exception: 
````
  File "C:\Anaconda\lib\site-packages\pandas\core\frame.py", line 1410, in to_csv
    formatter.save()
  File "C:\Anaconda\lib\site-packages\pandas\core\format.py", line 959, in save
    quoting=self.quoting)
  File "C:\Anaconda\lib\site-packages\pandas\core\common.py", line 1770, in __init__
    self.writer = csv.writer(self.queue, dialect=dialect, **kwds)
TypeError: "delimiter" must be an 1-character string
````
#### The workaround
-----------------------
If I change this to:
``` dataframe.to_csv(fname, sep=str(sep), index=index, float_format=float_format, **csvkwargs)``` everything works as you might expect. (I have observerved this error both when I pass in no encoding and when I pass in `encoding='Windows 1252'`)

#### What I would expect
------------------------------
When passed unicode, pandas should:
 * If encoding is set, and the unicode `sep` is one character long i.e. `len(sep)==1` (is it really necessary to have this one character restriction?), use the `sep` character passed in (encoded appropriately)
 * If encoding is not set, then check that the default encoding (I don't know whether this is ascii or platform dependent) of `sep` is only one character long

**Comment by gfyoung** [comment]
This issue persists in `master` using Python 2.7.12:
~~~python
>>> from __future__ import unicode_literals
>>> from pandas import DataFrame
>>>
>>> df = DataFrame([['a', 'A'], ['b', 'B']])
>>> df.to_csv(sep=',')
...
    self.writer = csv.writer(f, **writer_kwargs)
TypeError: "delimiter" must be string, not unicode
~~~
Note, that the error is raised when we instantiate the writer, which is a Python `csv` issue.  Thus, I would not consider this a `pandas` bug, but rather a compatibility issue with Python 2.x that is beyond our control.

The workaround of specifying `str(',')` still works:
~~~python
>>> df.to_csv(sep=str(','))
',0,1\n0,a,b\n1,A,B\n'
~~~
In short, I think this issue can be closed in the light of reason I provided above.

**Comment by jreback** [comment]
ok, I agree this is not a pandas bug and enough of an error message (though *maybe* could be an extended error message on what to do, though would have to catch this specific message and re-raise with a more informative one). I suppose that would be ok. closing in any event.

**Comment by salemy-m07** [comment]
hello some one can help me please 
![code png2](https://github.com/pandas-dev/pandas/assets/165857438/cc5e00a0-4738-4d10-8148-8281aa642f41)
This is my code and the File Data cant Download any data (.csv) to my file
I give This error...
[Error.txt](https://github.com/pandas-dev/pandas/files/14971204/Error.txt)
